File: src/main/scala/mffs/tile/TileCoercionDeriver.java
package mffs.tile;

import resonant.api.mffs.modules.IModule;
import com.google.common.io.ByteArrayDataInput;
import mffs.ModularForceFieldSystem;
import mffs.Settings;
import mffs.base.TileMFFSElectrical;
import mffs.FortronHelper;
import mffs.item.card.ItemCardFrequency;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraftforge.common.ForgeDirection;
import universalelectricity.api.CompatibilityModule;
import universalelectricity.api.energy.EnergyStorageHandler;

import java.io.IOException;
import java.util.EnumSet;

/**
 * A TileEntity that extract energy into Fortron.
 *
 * @author Calclavia
 */
public class TileCoercionDeriver extends TileMFFSElectrical
{
	public static final int FUEL_PROCESS_TIME = 10 * 20;
	public static final int MULTIPLE_PRODUCTION = 4;
	/**
	 * Ration from UE to Fortron. Multiply J by this value to convert to Fortron.
	 */
	public static final float UE_FORTRON_RATIO = 0.005f;
	public static final int ENERGY_LOSS = 1;
	public static final int SLOT_FREQUENCY = 0;
	public static final int SLOT_BATTERY = 1;

	public static final int SLOT_FUEL = 2;
	/**
	 * The amount of KiloWatts this machine uses.
	 */
	private static final int DEFAULT_WATTAGE = 5000000;
	public int processTime = 0;
	public boolean isInversed = false;

	public TileCoercionDeriver()
	{
		super();
		this.capacityBase = 30;
		this.startModuleIndex = 3;
		this.energy = new EnergyStorageHandler();
		updateEnergyInfo();
	}

	private void updateEnergyInfo()
	{
		this.energy.setCapacity(getWattage());
		this.energy.setMaxTransfer(getWattage() / 20);
	}

	@Override
	public void initiate()
	{
		super.initiate();
		updateEnergyInfo();
	}

	@Override
	public void updateEntity()
	{
		super.updateEntity();

		if (!worldObj.isRemote)
		{
			if (isActive())
			{
				if (isInversed && Settings.ENABLE_ELECTRICITY)
				{
					if (energy.getEnergy() < energy.getEnergyCapacity())
					{
						long withdrawnElectricity = (long) (requestFortron(getProductionRate() / 20, true) / UE_FORTRON_RATIO);
						// Inject electricity from Fortron.
						energy.receiveEnergy(withdrawnElectricity * ENERGY_LOSS, true);
					}

					recharge(getStackInSlot(SLOT_BATTERY));
					produce();
				}
				else
				{
					if (this.getFortronEnergy() < this.getFortronCapacity())
					{
						// Convert Electricity to Fortron
						this.discharge(this.getStackInSlot(SLOT_BATTERY));

						if (this.energy.checkExtract() || (!Settings.ENABLE_ELECTRICITY && this.isItemValidForSlot(SLOT_FUEL, this.getStackInSlot(SLOT_FUEL))))
						{
							// Fill Fortron
							this.fortronTank.fill(FortronHelper.getFortron(this.getProductionRate()), true);
							this.energy.extractEnergy();

							// Use fuel
							if (this.processTime == 0 && this.isItemValidForSlot(SLOT_FUEL, this.getStackInSlot(SLOT_FUEL)))
							{
								this.decrStackSize(SLOT_FUEL, 1);
								this.processTime = FUEL_PROCESS_TIME * Math.max(this.getModuleCount(ModularForceFieldSystem.itemModuleScale) / 20, 1);
							}

							if (this.processTime > 0)
							{
								// We are processing.
								this.processTime--;

								if (this.processTime < 1)
								{
									this.processTime = 0;
								}
							}
							else
							{
								this.processTime = 0;
							}
						}
					}
				}
			}
		}
		else if (this.isActive())
		{
			this.animation++;
		}
	}

	public long getWattage()
	{
		return (long) (DEFAULT_WATTAGE + (DEFAULT_WATTAGE * ((float) this.getModuleCount(ModularForceFieldSystem.itemModuleSpeed) / (float) 8)));
	}

	@Override
	public long onReceiveEnergy(ForgeDirection from, long receive, boolean doReceive)
	{
		if (!isInversed)
		{
			return super.onReceiveEnergy(from, receive, doReceive);
		}
		return receive;
	}

	@Override
	public long onExtractEnergy(ForgeDirection from, long extract, boolean doExtract)
	{
		if (isInversed)
		{
			return super.onExtractEnergy(from, extract, doExtract);
		}
		return 0;
	}

	@Override
	public void onInventoryChanged()
	{
		super.onInventoryChanged();
		updateEnergyInfo();
	}

	@Override
	public EnumSet<ForgeDirection> getOutputDirections()
	{
		return EnumSet.allOf(ForgeDirection.class);
	}

	/**
	 * @return The Fortron production rate per tick!
	 */
	public int getProductionRate()
	{
		if (this.isActive())
		{
			int production = (int) ((float) getWattage() / 20f * UE_FORTRON_RATIO * Settings.FORTRON_PRODUCTION_MULTIPLIER);

			if (this.processTime > 0)
			{
				production *= MULTIPLE_PRODUCTION;
			}

			return production;
		}

		return 0;
	}

	@Override
	public int getSizeInventory()
	{
		return 6;
	}

	public boolean canConsume()
	{
		if (this.isActive() && !this.isInversed)
		{
			return FortronHelper.getAmount(this.fortronTank) < this.fortronTank.getCapacity();
		}

		return false;
	}

	@Override
	public void onReceivePacket(int packetID, ByteArrayDataInput dataStream) throws IOException
	{
		super.onReceivePacket(packetID, dataStream);

		if (packetID == TilePacketType.TOGGLE_MODE.ordinal())
		{
			this.isInversed = !this.isInversed;
		}
	}

	@Override
	public void readFromNBT(NBTTagCompound nbt)
	{
		super.readFromNBT(nbt);
		this.processTime = nbt.getInteger("processTime");
		this.isInversed = nbt.getBoolean("isInversed");
	}

	@Override
	public void writeToNBT(NBTTagCompound nbt)
	{
		super.writeToNBT(nbt);
		nbt.setInteger("processTime", this.processTime);
		nbt.setBoolean("isInversed", this.isInversed);
	}

	@Override
	public boolean isItemValidForSlot(int slotID, ItemStack itemStack)
	{
		if (itemStack != null)
		{
			if (slotID >= this.startModuleIndex)
			{
				return itemStack.getItem() instanceof IModule;
			}

			switch (slotID)
			{
				case SLOT_FREQUENCY:
					return itemStack.getItem() instanceof ItemCardFrequency;
				case SLOT_BATTERY:
					return CompatibilityModule.isHandler(itemStack.getItem());
				case SLOT_FUEL:
					return itemStack.isItemEqual(new ItemStack(Item.dyePowder, 1, 4)) || itemStack.isItemEqual(new ItemStack(Item.netherQuartz));
			}
		}

		return false;
	}

	@Override
	public boolean canConnect(ForgeDirection direction, Object obj)
	{
		return true;
	}

}

File: src/main/scala/mffs/production/TileCoercionDeriver.scala
package mffs.production

import cpw.mods.fml.relauncher.{Side, SideOnly}
import io.netty.buffer.ByteBuf
import mffs.base.{TileModuleAcceptor, TilePacketType}
import mffs.item.card.ItemCardFrequency
import mffs.util.FortronUtility
import mffs.{Content, Settings}
import net.minecraft.client.renderer.RenderBlocks
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.init.Items
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraftforge.common.util.ForgeDirection
import resonant.api.mffs.modules.IModule
import resonant.lib.content.prefab.TElectric
import resonant.lib.network.ByteBufWrapper.ByteBufWrapper
import resonant.lib.network.discriminator.PacketType
import universalelectricity.api.UniversalClass
import universalelectricity.compatibility.Compatibility
import universalelectricity.core.transform.vector.Vector3

/**
 * A TileEntity that extract energy into Fortron.
 *
 * @author Calclavia
 */
object TileCoercionDeriver
{
  val fuelProcessTime = 10 * 20
  val productionMultiplier = 4

  /**
   * Ration from UE to Fortron. Multiply J by this value to convert to Fortron.
   */
  val ueToFortronRatio = 0.005f
  val energyConversionPercentage = 1

  val slotFrequency = 0
  val slotBattery = 1
  val slotFuel = 2

  /**
   * The amount of power (watts) this machine uses.
   */
  val power = 5000000
}

@UniversalClass
class TileCoercionDeriver extends TileModuleAcceptor with TEnergyBridge
{
  var processTime: Int = 0
  var isInversed: Boolean = false

  //Client
  var animationTween = 0f

  capacityBase = 30
  startModuleIndex = 3

  override def getSizeInventory = 6

  override def start()
  {
    super.start()
  }

  override def update()
  {
    super.update()

    if (!worldObj.isRemote)
    {
      if (isActive)
      {
        if (isInversed && Settings.enableElectricity)
        {
          val withdrawnElectricity = (requestFortron(productionRate / 20, true) / TileCoercionDeriver.ueToFortronRatio)
          electricNode.addEnergy(ForgeDirection.UNKNOWN, withdrawnElectricity * TileCoercionDeriver.energyConversionPercentage, true)

          recharge(getStackInSlot(TileCoercionDeriver.slotBattery))
        }
        else
        {
          if (getFortronEnergy < getFortronCapacity)
          {
            discharge(getStackInSlot(TileCoercionDeriver.slotBattery))
            val energy = electricNode.getEnergy(ForgeDirection.UNKNOWN)

            if (energy >= getPower || (!Settings.enableElectricity && isItemValidForSlot(TileCoercionDeriver.slotFuel, getStackInSlot(TileCoercionDeriver.slotFuel))))
            {
              fortronTank.fill(FortronUtility.getFortron(productionRate), true)
              electricNode.removeEnergy(ForgeDirection.UNKNOWN, getPower, true)

              if (processTime == 0 && isItemValidForSlot(TileCoercionDeriver.slotFuel, getStackInSlot(TileCoercionDeriver.slotFuel)))
              {
                decrStackSize(TileCoercionDeriver.slotFuel, 1)
                processTime = TileCoercionDeriver.fuelProcessTime * Math.max(this.getModuleCount(Content.moduleScale) / 20, 1)
              }

              if (processTime > 0)
              {
                processTime -= 1

                if (processTime < 1)
                {
                  processTime = 0
                }
              }
              else
              {
                processTime = 0
              }
            }
          }
        }
      }
    }
    else
    {
      /**
       * Handle animation
       */
      if (isActive)
      {
        animation += 1

        if (animationTween < 1)
          animationTween += 0.01f
      }
      else
      {
        if (animationTween > 0)
          animationTween -= 0.01f
      }
    }
  }

  def getPower: Double = TileCoercionDeriver.power + (TileCoercionDeriver.power * (getModuleCount(Content.moduleSpeed) / 8d))

  override def getVoltage = 1000D

  /**
   * @return The Fortron production rate per tick!
   */
  def productionRate: Int =
  {
    if (this.isActive)
    {
      var production = (getPower.asInstanceOf[Float] / 20f * TileCoercionDeriver.ueToFortronRatio * Settings.fortronProductionMultiplier).asInstanceOf[Int]

      if (processTime > 0)
      {
        production *= TileCoercionDeriver.productionMultiplier
      }

      return production
    }
    return 0
  }

  def canConsume: Boolean =
  {
    if (this.isActive && !this.isInversed)
    {
      return FortronUtility.getAmount(this.fortronTank) < this.fortronTank.getCapacity
    }
    return false
  }

  override def write(buf: ByteBuf, id: Int)
  {
    super.write(buf, id)

    if (id == TilePacketType.description.id)
    {
      buf <<< isInversed
      buf <<< processTime
    }
  }

  override def read(buf: ByteBuf, id: Int, player: EntityPlayer, packet: PacketType): Boolean =
  {
    super.read(buf, id, player, packet)

    if (world.isRemote)
    {
      if (id == TilePacketType.description.id)
      {
        isInversed = buf.readBoolean()
        processTime = buf.readInt()
        return true
      }
    }
    else
    {
      if (id == TilePacketType.toggleMoe.id)
      {
        isInversed = !isInversed
        return true
      }
    }

    return false
  }

  override def readFromNBT(nbt: NBTTagCompound)
  {
    super.readFromNBT(nbt)
    processTime = nbt.getInteger("processTime")
    isInversed = nbt.getBoolean("isInversed")
  }

  override def writeToNBT(nbt: NBTTagCompound)
  {
    super.writeToNBT(nbt)
    nbt.setInteger("processTime", processTime)
    nbt.setBoolean("isInversed", isInversed)
  }

  override def isItemValidForSlot(slotID: Int, itemStack: ItemStack): Boolean =
  {
    if (itemStack != null)
    {
      if (slotID >= this.startModuleIndex)
      {
        return itemStack.getItem.isInstanceOf[IModule]
      }
      slotID match
      {
        case TileCoercionDeriver.slotFrequency =>
          return itemStack.getItem.isInstanceOf[ItemCardFrequency]
        case TileCoercionDeriver.slotBattery =>
          return Compatibility.isHandler(itemStack.getItem)
        case TileCoercionDeriver.slotFuel =>
          return itemStack.isItemEqual(new ItemStack(Items.dye, 1, 4)) || itemStack.isItemEqual(new ItemStack(Items.quartz))
      }
    }
    return false
  }

  @SideOnly(Side.CLIENT)
  override def renderStatic(renderer: RenderBlocks, pos: Vector3, pass: Int): Boolean =
  {
    return false
  }

  @SideOnly(Side.CLIENT)
  override def renderDynamic(pos: Vector3, frame: Float, pass: Int)
  {
    RenderCoercionDeriver.render(this, pos.x, pos.y, pos.z, frame, isActive, false)
  }

  @SideOnly(Side.CLIENT)
  override def renderInventory(itemStack: ItemStack)
  {
    RenderCoercionDeriver.render(this, -0.5, -0.5, -0.5, 0, true, true)
  }
}